libraries/AP_UAVCAN/AP_UAVCAN_Server.cpp

```diff
@@ -0,0 +1,590 @@
+
+AP_UAVCAN_Server AP_UAVCAN_Server::_server(StorageAccess(StorageManager::StorageCANDNA));
+
+static uavcan::Publisher<uavcan::protocol::dynamic_node_id::Allocation>* allocation_pub_[MAX_NUMBER_OF_CAN_DRIVERS];
+
+//links the HereFlow messages to the backend
+void AP_UAVCAN_Server::subscribe_msgs(AP_UAVCAN* ap_uavcan)
+{
+    if (ap_uavcan == nullptr) {
+        return;
+    }
+
+    auto* node = ap_uavcan->get_node();
+    //Register Allocation Message Handler
+    uavcan::Subscriber<uavcan::protocol::dynamic_node_id::Allocation, AllocationCb> *AllocationListener;
+    AllocationListener = new uavcan::Subscriber<uavcan::protocol::dynamic_node_id::Allocation, AllocationCb>(*node);
+    const int alloc_listener_res = AllocationListener->start(AllocationCb(ap_uavcan, &handleAllocation));
+    if (alloc_listener_res < 0) {
+        AP_HAL::panic("Allocation Subscriber start problem\n\r");
+        return;
+    }
+    AllocationListener->allowAnonymousTransfers();
+
+    //Register Node Status Listener
+    uavcan::Subscriber<uavcan::protocol::NodeStatus, NodeStatusCb> *NodeStatusListener;
+    NodeStatusListener = new uavcan::Subscriber<uavcan::protocol::NodeStatus, NodeStatusCb>(*node);
+    const int nodestatus_listener_res = NodeStatusListener->start(NodeStatusCb(ap_uavcan, &handleNodeStatus));
+    if (nodestatus_listener_res < 0) {
+        AP_HAL::panic("NodeStatus Subscriber start problem\n\r");
+        return;
+    }
+}
+
+
+void AP_UAVCAN_Server::getHash(NodeData &node_data, const uint8_t unique_id[], const uint8_t &size) const
+{
+    uint64_t hash = FNV_1_OFFSET_BASIS_64;
+    hash_fnv_1a(size, unique_id, &hash);
+
+    // xor-folding per http://www.isthe.com/chongo/tech/comp/fnv/
+    hash = (hash>>56) ^ (hash&(((uint64_t)1<<56)-1));
+
+    // write it to ret
+    for (uint8_t i=0; i<6; i++) {
+        node_data.hwid_hash[i] = (hash >> (8*i)) & 0xff;
+    }
+}
+
+bool AP_UAVCAN_Server::readNodeData(NodeData &data, const uint8_t &node_id)
+{
+    if (node_id > 127) {
+        return false;
+    }
+    if (!storage_.read_block(&data, sizeof(occupation_mask_) + (node_id * sizeof(struct NodeData)), sizeof(struct NodeData))) {
+        server_state_ = STORAGE_FAILURE;
+        return false;
+    }
+    return true;
+}
+
+bool AP_UAVCAN_Server::writeNodeData(const NodeData &data, const uint8_t &node_id)
+{
+    if (node_id > 127) {
+        return false;
+    }
+    if (!storage_.write_block(sizeof(occupation_mask_) + (node_id * sizeof(struct NodeData)), 
+                             &data, sizeof(struct NodeData))) {
+        server_state_ = STORAGE_FAILURE;
+        return false;
+    }
+    return true;
+}
+
+bool AP_UAVCAN_Server::setOccupationMask(uint8_t bit)
+{
+    if (bit > 127) {
+        return false;
+    }
+
+    occupation_mask_[bit/8] |= (1 << (bit % 8));
+    //Write to storage
+    if (!storage_.write_block(0 ,occupation_mask_, sizeof(occupation_mask_) - 1)) {
```

Comment:
I don't think occupation_mask should be in persistent storage. If it is needed then build it at startup into memory, using our BitMask class